### PR TITLE
auto select acm cert 

### PR DESCRIPTION
**Pull request by szuecs**
If the ingress object does not provide zalando.org/aws-load-balancer-ssl-cert annotation, then we  automatically create/update/delete our LB resources.

### CODE REVIEW

File: aws/acm.go [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+		return
+	}
+
+	certDetails := make([]*acm.CertificateDetail, len(certList), len(certList))
+	for idx, o := range certList {
+		certInput := &acm.DescribeCertificateInput{CertificateArn: o.CertificateArn}
+		certDetail, err := cc.acmClient.DescribeCertificate(certInput)
+		if err != nil {
+			log.Printf("Could not get certificate details from AWS for ARN: %s, caused by: %v", o.CertificateArn, err)
+			return
```
Comment (lmineiro) [inline aws/acm.go]: Hmm, a `break` could be more appropriate. I can see how this could help with consistency but, is it really an issue? How different would it be from an outdated cache?

File: aws/acm.go [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+	return certList, nil
+}
+
+// FindBestMatchingCertifcate get all ACM certifcates and use a suffix
+// search best match opertion in order to find the best matching
+// certifcate ARN.
+//
+// We don't need to validate the Revocation here, because we only pull
+// ISSUED certificates.
+func FindBestMatchingCertifcate(certs []*acm.CertificateDetail, hostname string) (*acm.CertificateDetail, error) {
```
Comment (lmineiro) [inline aws/acm.go]: typo `FindBestMatchingCertifcate`
s/FindBestMatchingCertifcate/FindBestMatchingCertificate

File: aws/acm.go [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+		for _, cert := range resp.CertificateSummaryList {
+			certList = append(certList, cert)
+		}
+	}
+
+	return certList, nil
+}
+
+// FindBestMatchingCertifcate get all ACM certifcates and use a suffix
+// search best match opertion in order to find the best matching
```
Comment (lmineiro) [inline aws/acm.go]: typo `opertion `
s/opertion/operation

File: aws/acm.go [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+// certifcate ARN.
```
Comment (lmineiro) [inline aws/acm.go]: typo `certifcate`
s/certifcate/certificate

File: aws/acm.go [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+// ISSUED certificates.
+func FindBestMatchingCertifcate(certs []*acm.CertificateDetail, hostname string) (*acm.CertificateDetail, error) {
+	candidate := &acm.CertificateDetail{}
+	longestMatch := -1
+	sevenDaysInSec := int64(7 * 24 * 3600)
+
+	for _, cert := range certs {
+		now := time.Now().Unix()
+		notAfter := cert.NotAfter.Unix()
+		notBefore := cert.NotBefore.Unix()
```
Comment (lmineiro) [inline aws/acm.go]: Shouldn't these 3 statements be outside of the door loop? The issued dates are already set in the cert details. Also, this isn't supposed to take that long.

File: aws/acm.go [hunk 6]
```diff
@@ -0,0 +1,213 @@
...
+		// ignore invalid timeframes
+		if notAfter <= now || now <= notBefore {
```
Comment (lmineiro) [inline aws/acm.go]: Why not use the Time functions `After()` and `Before()`? Do we really care about equal?

File: aws/acm.go [hunk 7]
```diff
@@ -0,0 +1,213 @@
...
+			continue
+		}
+		altNames := append(cert.SubjectAlternativeNames, cert.DomainName)
+
+		for _, altName := range altNames {
+			if prefixGlob(aws.StringValue(altName), hostname) {
```
Comment (lmineiro) [inline aws/acm.go]: Extract `aws.StringValue(altName)`

File: aws/acm.go [hunk 8]
```diff
@@ -0,0 +1,213 @@
...
+// FindBestMatchingCertifcate get all ACM certifcates and use a suffix
+// search best match opertion in order to find the best matching
+// certifcate ARN.
+//
+// We don't need to validate the Revocation here, because we only pull
+// ISSUED certificates.
+func FindBestMatchingCertifcate(certs []*acm.CertificateDetail, hostname string) (*acm.CertificateDetail, error) {
+	candidate := &acm.CertificateDetail{}
+	longestMatch := -1
+	sevenDaysInSec := int64(7 * 24 * 3600)
```
Comment (lmineiro) [inline aws/acm.go]: This should be promoted to a documented constant
Does it stand for our eligibility cut off point?

File: aws/adapter.go [hunk 9]
```diff
@@ -67,24 +74,42 @@ func defaultConfigProvider() client.ConfigProvider {
...
 	return
 }
 
+func (a *Adapter) NewAcm(certUpdateInterval time.Duration) *certificateCache {
+	cc := NewCertCache(a.acm)
+	cc.InitCertCache(certUpdateInterval)
+	return cc
+}
+
+func (a *Adapter) GetCerts() []*acm.CertificateDetail {
```
Comment (lmineiro) [inline aws/adapter.go]: The adapter shouldn't export any AWS SDK specific types. Create a wrapper struct

File: aws/adapter.go [hunk 10]
```diff
@@ -67,24 +74,42 @@ func defaultConfigProvider() client.ConfigProvider {
...
-func NewAdapter(healthCheckPath string, healthCheckPort uint16) (adapter *Adapter, err error) {
+func NewAdapter(healthCheckPath string, healthCheckPort uint16, certUpdateInterval time.Duration) (adapter *Adapter, err error) {
 	p := configProvider()
 	adapter = &Adapter{
 		elbv2:           elbv2.New(p),
 		ec2:             ec2.New(p),
 		ec2metadata:     ec2metadata.New(p),
 		autoscaling:     autoscaling.New(p),
+		acm:             acm.New(p),
+		cc:              nil,
```
Comment (lmineiro) [inline aws/adapter.go]: not needed

File: worker.go [hunk 11]
```diff
@@ -60,19 +64,35 @@ func doWork(awsAdapter *aws.Adapter, kubeAdapter *kubernetes.Adapter) error {
...
 	}
 	return uniqueARNs
 }
 
+func findCertARNForIngress(awsAdapter *aws.Adapter, ingress *kubernetes.Ingress, acmCerts []*acm.CertificateDetail) (string, error) {
+	acmCert, err := awsAdapter.FindBestMatchingCertifcate(acmCerts, ingress.CertHostname())
+	if err != nil {
+		return "", err
+	}
+	ingress.SetCertificateARN(awssdk.StringValue(acmCert.CertificateArn))
```
Comment (lmineiro) [inline worker.go]: Extract `awssdk.StringValue(acmCert.CertificateArn)`

File: aws/acm.go (hunk 1 above)
Comment (szuecs) [inline aws/acm.go]: The problem is, that if AWS rate limited the call to the important certificate and your ingress use automatic selected certificates, then ingress-controller will delete this LB.

File: aws/acm.go (hunk 4 above)
Comment (szuecs) [inline aws/acm.go]: ty done

File: aws/acm.go (hunk 3 above)
Comment (szuecs) [inline aws/acm.go]: ty done

File: aws/acm.go (hunk 2 above)
Comment (szuecs) [inline aws/acm.go]: ty done

File: aws/acm.go (hunk 5 above)
Comment (szuecs) [inline aws/acm.go]: I did it for readability and I can only have now := ... outside of the loop, because the others are pointing to the iterator var.

File: aws/acm.go (hunk 7 above)
Comment (szuecs) [inline aws/acm.go]: fixed

File: aws/acm.go (hunk 6 above)
Comment (szuecs) [inline aws/acm.go]: I changed it according your suggestion and for the edge case I added a second to the test, such that the default NotAfter attribute is 7d+1s

File: aws/adapter.go (hunk 9 above)
Comment (szuecs) [inline aws/adapter.go]: true, unexported everything besides CertDetail (business object), FindBestMatchingCertificate, GetCachedCerts in acm.go and FindBestMatchingCertificate and GetCerts in adapter

File: aws/acm.go [hunk 12]
```diff
@@ -0,0 +1,240 @@
...
+	}
+
+	return result
+}
+
+// LsitCerts returns a list of acm Certificates filtered by
+// CertificateStatuses
+// https://docs.aws.amazon.com/acm/latest/APIReference/API_ListCertificates.html#API_ListCertificates_RequestSyntax
+func (cc *certificateCache) listCerts() ([]*acm.CertificateSummary, error) {
+	maxItems := aws.Int64(10)
```
Comment (tuxlife) [inline aws/acm.go]: Why did you set a `maxItems` value?

File: aws/acm.go [hunk 13]
```diff
@@ -0,0 +1,240 @@
...
+		notBefore := cert.NotBefore
+
+		// ignore invalid timeframes
+		if now.After(notAfter) || now.Before(notBefore) {
+			continue
+		}
+
+		for _, altName := range cert.AltNames {
+			if prefixGlob(altName, hostname) {
+				l := len(altName)
```
Comment (tuxlife) [inline aws/acm.go]: Please use a readable Name like `nameLength`. `1` and `l` looks to simulare

File: aws/acm.go (hunk 8 above)
Comment (szuecs) [inline aws/acm.go]: Yes and it should be fixed

File: aws/acm.go [hunk 14]
```diff
@@ -0,0 +1,244 @@
...
+func newCertCache(acmClient acmiface.ACMAPI) *certificateCache {
+	return &certificateCache{
+		acmClient:      acmClient,
+		acmCertSummary: make([]*acm.CertificateSummary, 0),
+		acmCertDetail:  make([]*acm.CertificateDetail, 0),
+	}
+}
+
+func (cc *certificateCache) initCertCache(certUpdateInterval time.Duration) {
+	go func() {
```
Comment (tuxlife) [inline aws/acm.go]: Can you run `cc.updateCertCache()` before start the Goroutine

File: worker.go (hunk 11 above)
Comment (szuecs) [inline worker.go]: done

File: aws/acm.go (hunk 14 above)
Comment (szuecs) [inline aws/acm.go]: done

File: aws/acm.go [hunk 15]
```diff
@@ -0,0 +1,240 @@
...
+func (cc *certificateCache) listCerts() ([]*acm.CertificateSummary, error) {
+	maxItems := aws.Int64(10)
+
+	params := &acm.ListCertificatesInput{
+		CertificateStatuses: []*string{
+			aws.String("ISSUED"), // Required
+		},
+		MaxItems: maxItems,
+	}
+	resp, err := cc.acmClient.ListCertificates(params)
```
Comment (szuecs) [inline aws/acm.go]: done

File: aws/acm.go (hunk 12 above)
Comment (szuecs) [inline aws/acm.go]: dropped

File: aws/acm.go (hunk 13 above)
Comment (szuecs) [inline aws/acm.go]: depends on your font use, but I changed for you :)

**Comment by szuecs** [comment]
I changed also the aws session to use maxRetries 3.

**Comment by szuecs** [comment]
All requested changes are now fixed

File: aws/adapter.go [hunk 16]
```diff
@@ -53,38 +60,56 @@ var (
...
+func (a *Adapter) newAcm() *certificateCache {
+	cc := newCertCache(a.acm)
+	return cc
+}
+
+func (a *Adapter) GetCerts() []*CertDetail {
+	return a.manifest.certificateCache.GetCachedCerts()
+}
+
+func (a *Adapter) FindBestMatchingCertificate(certs []*CertDetail, hostname string) (*CertDetail, error) {
```
Comment (lmineiro) [inline aws/adapter.go]: Missing godoc for exported func

File: aws/adapter.go [hunk 17]
```diff
@@ -53,38 +60,56 @@ var (
...
+
 	return
 }
 
+func (a *Adapter) newAcm() *certificateCache {
+	cc := newCertCache(a.acm)
+	return cc
+}
+
+func (a *Adapter) GetCerts() []*CertDetail {
```
Comment (lmineiro) [inline aws/adapter.go]: Missing godoc for exported func

File: kubernetes/adapter.go [hunk 18]
```diff
@@ -44,25 +49,45 @@ func (i *Ingress) String() string {
 	return fmt.Sprintf("%s/%s", i.namespace, i.name)
 }
 
-// Hostname returns the DNS hostname associated with the ingress
+// Hostname returns the DNS LoadBalancer hostname associated with the
+// ingress got from Kubernetes Status
```
Comment (lmineiro) [inline kubernetes/adapter.go]: s/got/gotten

Comment (lmineiro) [inline kubernetes/adapter.go]: or obtained, whichever you prefer

File: aws/acm.go [hunk 19]
```diff
@@ -115,40 +115,20 @@ func (cc *certificateCache) GetCachedCerts() []*CertDetail {
...
-			return nil, err
-		}
-
-		for _, cert := range resp.CertificateSummaryList {
+	err := cc.acmClient.ListCertificatesPages(params, func(page *acm.ListCertificatesOutput, lastPage bool) bool {
+		for _, cert := range page.CertificateSummaryList {
 			certList = append(certList, cert)
 		}
-	}
+		return lastPage
```
Comment (tuxlife) [inline aws/acm.go]: this is wrong.

The Return means, should I run again? You says NO...

Comment (szuecs) [inline aws/acm.go]: I would expect that AWS API makes sense in use and not that it sucks as it does. Thanks for catching this

File: aws/adapter.go (hunk 16 above)
Comment (szuecs) [inline aws/adapter.go]: done

File: aws/adapter.go (hunk 17 above)
Comment (szuecs) [inline aws/adapter.go]: done

File: kubernetes/adapter.go (hunk 18 above)
Comment (szuecs) [inline kubernetes/adapter.go]: fixed also the godoc above

File: aws/acm.go [hunk 20]
```diff
@@ -125,7 +125,7 @@ func (cc *certificateCache) listCerts() ([]*acm.CertificateSummary, error) {
 		for _, cert := range page.CertificateSummaryList {
 			certList = append(certList, cert)
 		}
-		return true // never stop iterator
+		return page.NextToken != nil
```
Comment (tuxlife) [inline aws/acm.go]: This is duplicated with https://github.com/aws/aws-sdk-go/blob/master/aws/request/request_pagination.go#L75

**Comment by szuecs** [comment]
:+1:

**Comment by tuxlife** [comment]
:+1:
